tools/linter/adapters/bazel_linter.py

```diff
@@ -0,0 +1,174 @@
+    are required
+    """
+    try:
+        # Use bazel to get the list of external dependencies in XML format
+        proc = subprocess.run(
+            [binary, "query", "kind(http_archive, //external:*)", "--output=xml"],
+            capture_output=True,
+        )
+    except OSError:
+        raise
+
+    stdout = str(proc.stdout, "utf-8").strip()
+    root = ET.fromstring(stdout)
+
+    checksums = {}
+    # Parse all the http_archive rules in the XML output
+    for rule in root.findall('.//rule[@class="http_archive"]'):
+        urls_node = rule.find('.//list[@name="urls"]')
+        if urls_node is None:
+            continue
+        urls = [n.get("value") for n in urls_node.findall(".//string")]
+
+        checksum_node = rule.find('.//string[@name="sha256"]')
+        if checksum_node is None:
+            continue
+        checksum = checksum_node.get("value")
+
+        if not checksum:
+            continue
+
+        checksums[checksum] = is_required_checksum(urls)
+
+    return checksums
+
+
+def check_bazel(
+    filename: str,
+    checksums: Dict[str, bool],
+) -> List[LintMessage]:
+    original = ""
+    replacement = ""
+
+    with open(filename) as f:
+        for line in f:
+            original += f"{line}"
+
+            m = SHA256_REGEX.match(line)
+            if m:
+                sha256 = m.group("sha256")
+
+                if not checksums.get(sha256, True):
+                    continue
+
+            replacement += f"{line}"
+
+        if original == replacement:
+            return []
+
+        return [
+            LintMessage(
+                path=filename,
+                line=None,
+                char=None,
+                code=LINTER_CODE,
+                severity=LintSeverity.ADVICE,
+                name="format",
+                original=original,
+                replacement=replacement,
+                description="Found redundant SHA checksums. Run `lintrunner -a` to apply this patch.",
+            )
+        ]
+
+
+def main() -> None:
+    parser = argparse.ArgumentParser(
+        description="A custom linter to detect redundant SHA checksums in Bazel",
+        fromfile_prefix_chars="@",
+    )
+    parser.add_argument(
+        "--binary",
+        required=True,
+        help="bazel binary path",
+    )
+    parser.add_argument(
+        "filenames",
+        nargs="+",
+        help="paths to lint",
+    )
+    args = parser.parse_args()
+
+    try:
+        checksums = get_checksums(args.binary)
+    except Exception as e:
+        err_msg = LintMessage(
+            path=None,
+            line=None,
+            char=None,
+            code=LINTER_CODE,
+            severity=LintSeverity.ERROR,
+            name="command-failed",
+            original=None,
+            replacement=None,
+            description=(f"Failed due to {e.__class__.__name__}:\n{e}"),
+        )
+        print(json.dumps(err_msg._asdict()), flush=True)
+        exit(0)
```

Comment:
Shouldn't this exit with a failure code?
```suggestion
        exit(1)
```

Reply:
It's lintrunner convention.  As long as the format failure is handle correctly, the code returns 0, for example https://github.com/pytorch/pytorch/blob/master/tools/linter/adapters/actionlint_linter.py#L134